### PR TITLE
Recurse submodules when cloning npm and yarn repos

### LINKED ISSUE #1166: yarn workspace with git submodule
In our setup some of the yarn workspaces live in a git submodule. It seems like dependabot runs yarn before (or without) pulling down the submodules. Portions of our `yarn.lock` come from packages in the submodule so this causes dependabot to delete thousands of line in each Bump PR. 

What do you think about adding an option to initialize the submodules before running yarn?

### LINKED ISSUE #5977: Dependabot corrupts `package-lock.json` when linking forked libraries as local git submodules
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Package ecosystem

npm

### Package manager version

8.11.0

### Language version

Nodejs@16.16.0

### Manifest location and content before the Dependabot update

https://github.com/swagger-api/swagger-editor/blob/2919f3d5b09fcfe46c29f8263075f265aae78116/package-lock.json

### dependabot.yml content

```yaml
version: 2
updates:
  - package-ecosystem: npm
    directory: "/"
    schedule:
      interval: daily
    commit-message:
      prefix: "chore"
      include: "scope"
    open-pull-requests-limit: 3
    ignore:
      # node-fetch must be synced manually
      - dependency-name: "node-fetch"
      - dependency-name: "release-it"
      - dependency-name: "@release-it/conventional-changelog"

  - package-ecosystem: "docker"
    # Look for a `Dockerfile` in the `root` directory
    directory: "/"
    # Check for updates once a week
    schedule:
      interval: "weekly"

  - package-ecosystem: "github-actions"
    directory: "/"
    schedule:
      interval: "weekly"

  ######################
  # SwaggerEditor@next #
  ######################

  - package-ecosystem: npm
    directory: "/"
    schedule:
      interval: daily
    target-branch: "next"
    labels:
      - "swagger-editor@next"
      - "cat: dependencies"
      - "javascript"
    commit-message:
      prefix: "chore"
      include: "scope"
    open-pull-requests-limit: 3
    ignore:
      # do not update react-scripts as we use forked version of Create React App via git submodule
      - dependency-name: "create-react-app/packages/react-scripts"
      - dependency-name: "react-scripts"

  - package-ecosystem: "docker"
    # Look for a `Dockerfile` in the `root` directory
    target-branch: "next"
    directory: "/"
    # Check for updates once a week
    schedule:
      interval: "weekly"
    labels:
      - "swagger-editor@next"
      - "cat: dependencies"
      - "cat: docker"
      - "docker"

  - package-ecosystem: "github-actions"
    target-branch: "next"
    directory: "/"
    schedule:
      interval: "weekly"
    labels:
      - "swagger-editor@next"
      - "cat: dependencies"
      - "github_actions"
```

### Updated dependency

@asyncapi/parser from 1.17.0 to 1.17.1

### What you expected to see, versus what you actually saw

I expected to see that dependabot updates @asyncapi/parser@1.17.0 to @asyncapi/parser@1.17.1.  but dependabot updates also our forked Create React App (specifically `react-scripts`). We use forked Create React App as a git submodule and use the npm local link to insert it into `devDependencies`.

```json
"react-scripts": "file:./create-react-app/packages/react-scripts",
```

This worked like a charm, but last week, dependabot started to issue all PRs in way that it corrupts the `package-lock.json` and npm is unable to install the packages. I've also added some ignores for dependabot but, it had no effect.

```yaml
    ignore:
      # do not update react-scripts as we use forked version of Create React App via git submodule
      - dependency-name: "create-react-app/packages/react-scripts"
      - dependency-name: "react-scripts"
```

image

### Native package manager behavior

By doing update locally, `react-scripts` is not updated in any way.

### Images of the diff or a link to the PR, issue, or logs

https://github.com/swagger-api/swagger-editor/pull/3581

### Smallest manifest that reproduces the issue

_No response_

### LINKED ISSUE #7441: Issue with npm lock file v3 and sub modules: "npm ERR! `npm ci` can only install packages when your package.json and package-lock.json or npm-shrinkwrap.json are in sync. Please update your lock file with `npm install` before continuing."
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Package ecosystem

npm

### Package manager version

NPM: `9.5.1`

### Language version

Node: `18.16.0`

### Manifest location and content before the Dependabot update

- `./package.json`
- `./package-lock.json`

### dependabot.yml content

```yml
name: Dependabot Test

on:
  pull_request:
    paths:
      - 'package.json'
      - 'package-lock.json'

jobs:
  dependabot:
    runs-on: ubuntu-latest

    if: ${{ github.actor == 'dependabot[bot]' }}

    steps:
      - uses: actions/checkout@v3
        with:
          token: ${{ secrets.PERSONAL_ACCESS_TOKEN_GITHUB_TOKEN }}
          submodules: 'recursive'

      - uses: actions/setup-node@v3
        with:
          node-version: '18.x'
          cache: 'npm'

      - name: Setup SSH
        uses: webfactory/ssh-agent@v0.5.4
        with:
          ssh-private-key: ${{ secrets.SSH_PRIVATE_KEY }}

      - name: 'Installing Packages'
        run: npm ci
        env:
          GITHUB_TOKEN: ${{ secrets.PERSONAL_ACCESS_TOKEN_GITHUB_TOKEN }}

      - name: Run Lint
        run: npm run lint:dry

      - name: Run Build
        run: npm run build:all:noemit

      - name: Run Tests
        run: npm run test
```

### Updated dependency

_No response_

### What you expected to see, versus what you actually saw

npm packages should install successfully.

## What I Saw

`api-shared` is the submodule/package we're using.

```
npm ERR! code EUSAGE
npm ERR! 
npm ERR! `npm ci` can only install packages when your package.json and package-lock.json or npm-shrinkwrap.json are in sync. Please update your lock file with `npm install` before continuing.
npm ERR! 
npm ERR! Missing: api-shared@0.0.1 from lock file
npm ERR! Missing: @prisma/client@4.15.0 from lock file
npm ERR! Missing: @types/node@18.16.18 from lock file
... list continues for everything within the submodule's dependencies
```

### Native package manager behavior

_No response_

### Images of the diff or a link to the PR, issue, or logs

_No response_

### Smallest manifest that reproduces the issue

These have been provided above in the form of the dependabot build/test yml

### Related Issues

- #6270 (has been closed, but doesn't seem resolved?)

### PR DESCRIPTION
**Pull request by deivid-rodriguez**
This allows unrelated updates to not remove submodule dependency information from lockfiles due to the files being missing.

In addition to enabling cloning submodules, I also started ignoring updates for dependencies defined in the submodules themselves since they would get the submodule dirty and would not be mergeable.

Fixes #1166.
Fixes #5977.
Fixes #7441.

### CODE REVIEW

**Comment by kleinfreund** [comment]
As far as I can tell, the tests fail as expected because one asserts for the file fetcher class to not have the additional method `recurse_submodules_when_cloning?` which is added here.

**Comment by deivid-rodriguez** [comment]
Yeah, I think I just need to make the new method private and should be fine after that. But I'd like to look into getting this fix tested too.

**Comment by deivid-rodriguez** [comment]
I fixed the failing spec and confirmed that the problem still exists, and that this PR fixes it.

A test looks a bit tricky, and I don't want to block this fix on it, so I'll set this as ready.

**Review by landongrindheim — APPROVED** [review]

File: npm_and_yarn/lib/dependabot/npm_and_yarn/file_parser.rb [hunk 1]
```diff
@@ -334,6 +340,10 @@ def url_for_relevant_cred(resolved_url)
         resolved_url.gsub(/#{Regexp.quote(reg)}.*/, "") + reg
       end
 
+      def submodule_package_files
+        @submodule_package_files ||= package_files.select { |f| f.type == "submodule" }
```
Comment (abdulapopoola) [inline npm_and_yarn/lib/dependabot/npm_and_yarn/file_parser.rb:344]: is there any way to have test coverage? So we can catch regressions in the future easily

File: common/lib/dependabot/file_fetchers/base.rb [hunk 2]
```diff
@@ -684,6 +693,17 @@ def decode_binary_string(str)
...
       end
+
+      def find_submodules(path)
+        SharedHelpers.run_shell_command(
+          <<~CMD
+            git -C #{path} ls-files --stage
+          CMD
+        ).split("\n").filter_map do |line|
+          info = line.split
+          next info.last if info.first == "160000"
```
Comment (abdulapopoola) [inline common/lib/dependabot/file_fetchers/base.rb:704]: what is 160000? Is it a special number?

Comment (deivid-rodriguez) [reply common/lib/dependabot/file_fetchers/base.rb:704]: Yeah, it's the permissions git assigns to a submodule. I'll add comments for this.

File: npm_and_yarn/lib/dependabot/npm_and_yarn/file_parser.rb (hunk 1 above)
Comment (deivid-rodriguez) [reply npm_and_yarn/lib/dependabot/npm_and_yarn/file_parser.rb:344]: I think a smoke test would be best, I can work on it!

File: common/lib/dependabot/file_fetchers/base.rb (hunk 2 above)
Comment (deivid-rodriguez) [reply common/lib/dependabot/file_fetchers/base.rb]: I ended up adding https://github.com/dependabot/dependabot-core/pull/6718/commits/e2f040f7df958eff20588d5fb460a1a03fbf32a8, so no more magic number!

File: npm_and_yarn/lib/dependabot/npm_and_yarn/file_parser.rb (hunk 1 above)
Comment (deivid-rodriguez) [reply npm_and_yarn/lib/dependabot/npm_and_yarn/file_parser.rb]: A smoke test turned out to be tricky, so I added one regular spec to cover this, and corrected a bug it found.

**Comment by deivid-rodriguez** [comment]
I added a test here, and also found two more issues that I think this PR will fix!

**Review by abdulapopoola — APPROVED** [review]

**Pull request merged**